### PR TITLE
RFQ Relayer: add scroll rebalancing

### PR DESCRIPTION
**Pull request by dwasse**
## Summary by CodeRabbit

- **New Features**
	- Introduced support for new rebalance methods, including the `Scroll` method for enhanced cross-chain interactions.
	- Implemented broader retrieval of rebalance methods, allowing for dynamic management of multiple rebalance strategies.
	- Added approval logic for `L1GatewayRouter` and `L2GatewayRouter` contracts.

- **Enhancements**
	- Improved error handling and messaging for rebalance management, ensuring clarity when no valid rebalance method is found.
	- Enhanced the flexibility of token configurations by allowing multiple rebalance methods to be specified.
	- Streamlined address retrieval processes for improved type safety and reduced errors.
	- Added a new utility package to encapsulate reusable functions for the RFQ service.

### CODE REVIEW

File: services/rfq/relayer/inventory/manager.go [hunk 1]
```diff
@@ -307,45 +304,78 @@ func (i *inventoryManagerImpl) ApproveAllTokens(ctx context.Context) error {
...
+					return fmt.Errorf("could not get L1Gateway contract: %w", err)
+				}
+				contractAddr, err := contract.EthGateway(&bind.CallOpts{Context: ctx})
+				if err != nil {
+					return fmt.Errorf("could not get L1ETHGateway address: %w", err)
+				}
+				err = i.approve(ctx, tokenAddr, contractAddr, backendClient)
+				if err != nil {
+					return fmt.Errorf("could not approve L1ETHGateway contract: %w", err)
+				}
```
Comment (ChiTimesChi) [inline services/rfq/relayer/inventory/manager.go:347]: @dwasse Don't think we need to approve ETHGateway contract, only ERC20Gateway uses token approvals

File: services/rfq/relayer/inventory/manager.go [hunk 2]
```diff
@@ -307,45 +304,78 @@ func (i *inventoryManagerImpl) ApproveAllTokens(ctx context.Context) error {
...
-				err = i.approve(ctx, tokenAddr, common.HexToAddress(contractAddr), backendClient)
+				contractAddr, err := contract.EthGateway(&bind.CallOpts{Context: ctx})
 				if err != nil {
-					return fmt.Errorf("could not approve TokenMessenger contract: %w", err)
+					return fmt.Errorf("could not get L2ETHGateway address: %w", err)
+				}
+				err = i.approve(ctx, tokenAddr, contractAddr, backendClient)
+				if err != nil {
+					return fmt.Errorf("could not approve L2ETHGateway contract: %w", err)
+				}
```
Comment (ChiTimesChi) [inline services/rfq/relayer/inventory/manager.go:371]: Same as above re: EthGateway approvals

File: services/rfq/contracts/l2gateway/parser.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	}
+
+	return &parserImpl{filterer: parser}, nil
+}
+
+// nolint: cyclop
+func (p parserImpl) ParseEvent(log ethTypes.Log) (_ EventType, event interface{}, ok bool) {
+	// return an unknown event to avoid cases where user failed to check the event type
+	// make it high enough to make it obvious (we start iotas at +1, see uber style guide for details)
+	noOpEvent := EventType(len(topicMap()) + 1)
```
[the remark this answers is not in the document]
Comment (trajan0x) [reply services/rfq/contracts/l2gateway/parser.go:50]: not doable

File: services/rfq/relayer/inventory/manager.go [hunk 4]
```diff
@@ -307,45 +304,62 @@ func (i *inventoryManagerImpl) ApproveAllTokens(ctx context.Context) error {
 			return fmt.Errorf("could not get chain client: %w", err)
 		}
 
-		for address, token := range tokenMap {
-			// approve RFQ contract.
+		for tokenAddr := range tokenMap {
```
Comment (trajan0x) [inline services/rfq/relayer/inventory/manager.go:307]: just a note: this removes batching for allowances

**Review by trajan0x — APPROVED** [review]

**Pull request merged**